### ISSUE TITLE
Suite

### ISSUE THREAD

**Issue by rcharb1** [rcharb1/optimiseur-coupe#1]
Je suis ici

**Comment by pierre-H** [comment]
Génial

**Comment by rcharb1** [comment]
parfait! Maintenant, pour syncer ce fork sur le repository parent, je fais un pull request ?

**Comment by pierre-H** [comment]
oui

**Comment by rcharb1** [comment]
Xavier ne recevra pas de mails dans ce fork, puisque j'ai dû ajouter les collaborateurs que je veux

**Comment by pierre-H** [comment]
Ici : https://github.com/rcharb1/optimiseur-coupe il y a écrit 3 collaborators

**Comment by rcharb1** [comment]
et ici, chez moi, on n'est que 2... https://github.com/rcharb1/optimiseur-coupe/settings/collaboration

**Comment by pierre-H** [comment]
Page 404.
Combien vois tu sur cette page https://github.com/rcharb1/optimiseur-coupe ?

**Comment by rcharb1** [comment]
404 parceque les settings de ce fork ne sont accessibles qu'à moi...

**Comment by pierre-H** [comment]
oui

**Comment by rcharb1** [comment]
effectivement,il est écrit 3 contributors, mais jene suis pas sûre que ça veuille dire 3 collaborateurs, peut-être 3 personnes qui ont participé aux différents commits, ce qui est vrai...

**Comment by rcharb1** [comment]
on va faire un essai,je vais t'enlever comme collaborateur dans ma page de settings, écrire un message danscet issue, et on verra dans la branche Pierre-H si tu as reçu un mail de notification du message. C'est OK ?

**Comment by pierre-H** [comment]
ok

**Comment by rcharb1** [comment]
voici un message qui ne devrait parvenir à personne par mail.
Si c'est le cas, merci de bien vouloir le faireparvenir à son propriétaire etc etc.

**Comment by pierre-H** [comment]
Je l'ai reçu par mail

**Comment by rcharb1** [comment]
ah...

**Comment by pierre-H** [comment]
Peut être que collaborateurs et contributeurs ce n'est pas la même chose...

**Comment by rcharb1** [comment]
oui, je pense que la différence est là. Mais dans ce cas, jusqu'à ce qu'il fasse un unwatch, Xavier recevra tout, c'est ça ?

**Comment by rcharb1** [comment]
En tout cas, je ne pense pas qu'on a la possibilité de changer ce comportement...

**Comment by rcharb1** [comment]
avant de parler du projet proprement dit, j'ai eu des difficultés avec ma pull request qui me dit qu'il n'y a rien de différent entre les deux repos...

**Comment by pierre-H** [comment]
Normal

**Comment by pierre-H** [comment]
Il n'y rien de différent.

**Comment by rcharb1** [comment]
sauf 97 commits

**Comment by rcharb1** [comment]
en plus...

**Comment by pierre-H** [comment]
Le code est le même

**Comment by rcharb1** [comment]
donc il faut un merge ?

**Comment by pierre-H** [comment]
Pourquoi ? Le code est le même.
Cela nous est égal non ?

**Comment by rcharb1** [comment]
donc il est à la même version ? Alors c'est parfait, désolée pour ma lenteur...

**Comment by rcharb1** [comment]
Alors, abordons le sujet à proprement parler...

**Comment by pierre-H** [comment]
Oui. Juste concernant les disponibilités, en journée je suis au stage donc pas très dispo...

**Comment by rcharb1** [comment]
pas de problème, ça me va le soir, sauf demain soir !

**Comment by pierre-H** [comment]
Idem :)

**Comment by pierre-H** [comment]
Alors, concernant le projet... Il y aura le serveur qui accepte les résultats du formulaire et qui renvoie le résultat du moteur

**Comment by pierre-H** [comment]
La question concernant celui-ci est : faut-il intégrer une gui à ce moteur ou non...

**Comment by rcharb1** [comment]
Donc, pour résumer, on a : 
1. L'établissement d'une connexion avec le code gentiment fourni par M. Mehat
2. Il faut discuter d'une manière de transmettre les requêtes du client sur la socket.
3. Le moteur calcule comme il sait si bien le faire
4. De même, discuter de la manière de rendre les réponses au client.

**Comment by rcharb1** [comment]
pardon...

**Comment by pierre-H** [comment]
Pas de problème :)
Je suis d'accord avec les points.

**Comment by pierre-H** [comment]
Pour le  2e point  , on va devoir comme indiqué dans l'énoncé faire passer les données via un stringe

**Comment by rcharb1** [comment]
pour l'instant, dans le code de M. Mehat, seuls des strings sont transmis par la socket, n'est-ce pas ?

**Comment by pierre-H** [comment]
Oui

**Comment by rcharb1** [comment]
ah, donc il va falloir des fonctions de type eval...

**Comment by pierre-H** [comment]
pourquoi eval ? plutôt des méthodes permettant de couper des string et d'évaluer les nombres pour les transformer en double

**Comment by pierre-H** [comment]
On pourrait utiliser la fonction [split](http://doc.qt.io/qt-5/qstring.html#split) de QString qui permet de couper une chaîne de caractère

**Comment by rcharb1** [comment]
donc si dans les spécifications, les Lists doivent être représentées par (1, 5., 8.2), il faut une fonction qui reconstitue l'objet List correspondant à la sortie de la socket côté serveur, c'est ça ?

**Comment by pierre-H** [comment]
Et transformer un QString en double c'est très facile aussi (je pense l'avoir déjà fait pour l'optimiseur de coupe)

**Comment by rcharb1** [comment]
Mais il faut établir des spécifications très précises sur la forme que doit prendre chaque donnée...

**Comment by pierre-H** [comment]
Oui tout à fait

**Comment by rcharb1** [comment]
Tout va se faire dans un seul échange ou entre deux données il faut un ack du serveur ?

**Comment by pierre-H** [comment]
Je propose que l'on fasse la liste des choses à faire/décider et après choisir ensemble la façon dont on représentera les données

**Comment by pierre-H** [comment]
Cad ?

**Comment by rcharb1** [comment]
acknowlegment : j'ai bien reçu la liste des tronçons, envoie-moi maintenant celle des barres

**Comment by pierre-H** [comment]
Comme ça je dirais que ce n'est pas la peine, mais par sécurité on devrait peut être le faire. Qu'en penses tu ?

**Comment by rcharb1** [comment]
Ou bien un code que la liste des tronçons doit être séparée de la liste des barres par la chaîne de caractères "taratata" et qu'ainsi le serveur reconstitue le code...

**Comment by rcharb1** [comment]
Je pense que la deuxième solution est moins lourde...

**Comment by pierre-H** [comment]
Oui, mais dangereuse si il y a une coupure de connexion non ?
Mais je pense aussi qu'il ne faut pas trop se prendre la tête...
J'hésite...

**Comment by rcharb1** [comment]
Si coupure de connexion, remarque qu'il n'a pas reçu un caractère de fin défini d'avance et ne travaille donc pas dessus car détecte qu'il y a eu un pb... Checksum en quelque sorte...

**Comment by rcharb1** [comment]
C'est une solution ?

**Comment by pierre-H** [comment]
Alors on fait confiance au protocole ip ?

**Comment by rcharb1** [comment]
oui, mais en lui donnant une sécurité supplémentaire sous forme de string de fin défini d'avance, comme quoi le message a pu être transmis -> la fin... Si tout est arrivé, je ne crois pas qu'il soit nécessaire de le vérifier

**Comment by pierre-H** [comment]
Alors tu veux que l'on mette nous même en place un checksum ?

**Comment by pierre-H** [comment]
Autant utilisé un protocole qui prend en charge cela

**Comment by rcharb1** [comment]
un simili, juste une chaîne de fin dont on teste la présence,non ?

**Comment by pierre-H** [comment]
Ah oui on peut aussi. Je suis partant pour cette solution alors

**Comment by rcharb1** [comment]
ok.

**Comment by rcharb1** [comment]
Maintenant, gui ou non ?

**Comment by rcharb1** [comment]
Elle se fait de toutes les façons du côté user, n'est-ce pas ?

**Comment by pierre-H** [comment]
Pour le serveur, j'hésite. Le seul intérêt de la gui pour le serveur serait d'indiquer dans une fenêtre le port. C'est tout... Non ?

**Comment by rcharb1** [comment]
oui, je pense qu'il n'y a pas besoin de gui pour le serveur...

**Comment by rcharb1** [comment]
quelqu'un du côté serveur, c'est un pro, non ? ;-)

**Comment by pierre-H** [comment]
Ok, alors pour le moment pas de gui pour le serveur.
Un pro ? Qui ? L'utilisateur ? Si oui, je ne vois pas ce que cela change ...

**Comment by rcharb1** [comment]
Celui qui fait marcher le serveur... mais c'était une blague pour nous dispenser de gui! Bref

**Comment by pierre-H** [comment]
Ah ok :)

**Comment by rcharb1** [comment]
donc si on résume ce que le client doit envoyer : ...

**Comment by rcharb1** [comment]
Mes_barres_selon_une_syntaxe_definie      Séparation_entre_barres_et_tronçons     Mes_trnçons         Separ               Ma_Perte_A_La_Coupe

**Comment by rcharb1** [comment]
... fin d'envoi.

**Comment by rcharb1** [comment]
J'en ai oublié ?

**Comment by rcharb1** [comment]
Le serveur récupère la prose, la convertit en structures de données intelligibles, et renvoie : ...

**Comment by rcharb1** [comment]
Barre_0_:_tronçons Separ ... Barre_n_:_tronçons Separ Rendement_:_...%   Fin d'envoi

**Comment by pierre-H** [comment]
Oui. Je verrai les choses ainsi : 
`"START;[[20,3],[5,1],[6,1]];[[15,2],[4,2]];0;END"`

**Comment by rcharb1** [comment]
oui, pour le client, ça paraît parfait

**Comment by pierre-H** [comment]
Pour : 
3 barres de 20 mm
1 barre de 5 mm
1 barre de 6 mm

Tronçons désirés : 
2 barres de 15 mm
2 barres de 4 mm

0 pour la lame

**Comment by rcharb1** [comment]
et pour le serveur : ...

**Comment by rcharb1** [comment]
`"START;[[20:3,5.5,4.9,4.2];[80:60.5,8.3]];50;END"`

**Comment by pierre-H** [comment]
Je ne comprend pas comment interpréter ça : 
`"START;[[20:3,5.5,4.9,4.2];[80:60.5,8.3]];50;END"`

**Comment by rcharb1** [comment]
pour une barre de 20 avec les tronçons de 3 5.5 4.9 4.2
et une barre de 80 avec 60.5 et 8.3 dessus et un rendement totalde 50%

**Comment by rcharb1** [comment]
Du coup, deux travaux à entreprendre dans ce sens : 
1. La Gui qui génère la chaîne
2. Le serveur qui la traduit

**Comment by pierre-H** [comment]
oui, mais ne faut-il pas le rendement pour chaque tronçon ?
(c'est aussi l'ocassion de modifier le calcul du rendement moyen tel que l'a proposé gilles bernard non ?)

**Comment by rcharb1** [comment]
et au retour de poste, l'inverse :  le serveur qui génère la chaîne et l'interface qui la traduit avant de l'afficher

**Comment by pierre-H** [comment]
Générer et traduire la chaîne je pense que c'est très facile

**Comment by rcharb1** [comment]
C'est vrai... Il faut qu'on réfléchisse à cette histoire de rendement car perso, je n'ai pas encore trouvé de solution plausible...

**Comment by rcharb1** [comment]
du coup, à part la génération et la traduction, quel est réellement le travail de cet ajout ?

**Comment by pierre-H** [comment]
La partie réseau :)

**Comment by pierre-H** [comment]
Parce que là il faut se mettre d'accord sur les outils utilisés : quelle librairie choisir ! (encore cette question :) )

**Comment by rcharb1** [comment]
mais JM a dit qu'on peut utiliser ses fichiers de création de socket... donc on a la base ?

**Comment by pierre-H** [comment]
Qui sont en C... nous on est en C++

**Comment by rcharb1** [comment]
Ah, on va faire les choses un peu plus loin de la machine, ça me réjouit pas mal !

**Comment by rcharb1** [comment]
Qu'est-ce qu'il y a comme librairies ? Je cherche en même temps

**Comment by pierre-H** [comment]
> Ah, on va faire les choses un peu plus loin de la machine

cad ?

**Comment by rcharb1** [comment]
L'utilisation de certaines librairies ne définissent-elles pas des fonctions qui simplifient l'établissement d'une socket par rapport au code de JM ?

**Comment by pierre-H** [comment]
Au choix : la librairie standard ou celle de [Qt](https://openclassrooms.com/courses/programmez-avec-le-langage-c/communiquer-en-reseau-avec-son-programme)

**Comment by pierre-H** [comment]
Ben, si on utilise Qt, peut-être. Si c'est la librairie standard je ne pense pas.
La question est qu'est-ce qu'on a le droit de faire :)
Sachant que j'ai l'impression que cela aurait été mieux pour Gilles Bernard de ne pas mélanger les librairies...

**Comment by rcharb1** [comment]
on a de toutes les façons déjà mélangé... Jepense qu'on peut prendre Qt

**Comment by rcharb1** [comment]
Il faut que je me documente dessus

**Comment by pierre-H** [comment]
Moi aussi, je n'ai pas fait cette partie dans Qt, mais le cours dans OpenClassRoom à l'air bien...

**Comment by pierre-H** [comment]
Je relis de prêt les instructions de JM...

**Comment by rcharb1** [comment]
Donc, à part les fonctions de conversion string -> str de données et retour, il faut établir la connexion comme le fait JM, mais en C++, c'est ça ?

**Comment by pierre-H** [comment]
Oui

**Comment by pierre-H** [comment]
Pardon, avant de confirmer cela, je relis les instructions précises de JM...

**Comment by pierre-H** [comment]
ça à l'aire trop facile...

**Comment by rcharb1** [comment]
Sur le papier, ça a l'air, mais en réalité, je crois que ça va quand même prendre du temps... Pour moi, ça m'évoque un peu les bébés nageurs...

**Comment by rcharb1** [comment]
bon, mais il y a aussi la version plus compliquée...

**Comment by pierre-H** [comment]
Comment faire : `13.1.2 Version plus avancée` tant qu'on y est ?

**Comment by rcharb1** [comment]
Que le serveur puisse avoir une espèce de compte séparé pour chaque utilisateur, et puisse répondreà plusieurs sans semélanger les pinceaux, non ?

**Comment by pierre-H** [comment]
Oui

**Comment by rcharb1** [comment]
C'est quand même pas mal plus compliqué... Comment envisager ça ?

**Comment by rcharb1** [comment]
(Je viens de me souvenir que j'ai aussi quelque chose jeudi soir et que je ne pourrai donc pas travailler (mais le matin, je peux, donc si on peut se fixer une première partie du boulot à faire pour chacun en plus de lire le cours de openclassroom, ceserait bien...)

**Comment by rcharb1** [comment]
...) erreur de syntaxe : parenthèse ouvrante sans fermante !

**Comment by rcharb1** [comment]
Y a quelqu'un ?

**Comment by pierre-H** [comment]
Oui, je réfléchis...

**Comment by rcharb1** [comment]
pardon...

**Comment by rcharb1** [comment]
Je suis désolée,mais  ne vais pas tarder à y aller (1 heure de plus en Israel et une dernière nuit très courte...) Je propose qu'on se réunisse de nouveau dimanche soir ou lundi soir et qu'on discute de ce projet version compliquée. Pour ce qui est de ce qu'on a discuté, je propose qu'on se documente chacun de son côté et qu'on fasse chacun les deux fonctions de conversion l'un du côté client l'autre du côté serveur. Qu'en penses-tu ?

**Comment by pierre-H** [comment]
Ah oui désolé.
Ok pour tout ! Lundi soir pour moi c'est mieux.

**Comment by rcharb1** [comment]
TB, pour lundi soir, encore désolé, et disons que je m'occupe des deux fonctions côté serveur ? Et toi, côté client ? (au hasard)

**Comment by pierre-H** [comment]
Oui ok

**Comment by rcharb1** [comment]
Merci beaucoup, je crois qu'on a bien avancé ! Bonne nuit !

**Comment by pierre-H** [comment]
Je pense aussi.
Bonne nuit !
